core/common/src/main/java/alluxio/Configuration.java

```diff
@@ -573,5 +594,29 @@ public static void validate() {
     checkTieredLocality();
   }
 
+  /**
+   * Gets the configuration of a given scope.
+   *
+   * @param scope the property key scope
+   * @return a list of configurations inside the property scope
+   */
+  public static List<ConfigProperty> getConfiguration(Scope scope) {
+    List<ConfigProperty> list = new ArrayList<>();
+    for (Map.Entry<String, String> entry : toMap().entrySet()) {
+      PropertyKey key = PropertyKey.fromString(entry.getKey());
+      if (key.getScope().contains(scope)) {
+        ConfigProperty configProperty = new ConfigProperty()
+            .setName(key.getName()).setSource(getFormattedSource(key));
+        try {
+          configProperty.setValue(get(key));
+        } catch (RuntimeException e) {
+          configProperty.setValue(VALUE_NOT_SET_STRING);
```

Comment:
what if a property is set to the value "NOT_SET"? There should be a way for callers of this API to tell whether a property is set or not.

Reply:
What if we use an instruct to represent value include boolean isSet and String value？

Reply:
can you elaborate? How do we implement isSet?

Reply:
Actually, I prefer that if the value can be get via Configuration.get(key), we set the value of that ConfigProperty and if the  Configuration.get(key) throws runtime exception, we do nothing (do not set value for the ConfigProperty).  We add comments on the thrift file like what we did for zookeeperAddress and said that null unset value means the key has null default value.

We do not use concurrentHashMap in ServerConfigurationStore and ServerConfigurationChecker. The normal maps and lists support null value.

This way will also work.

Reply:
We can use `containsKey` to avoid generating and catching a `RuntimeException`.

What will happen if a property is required to be consistent across masters, and one master sets it to a `null` value while another sets to to non-null? How will we detect that inconsistency?

Reply:
containsKey will work.
```
        if (containsKey(key)) {
          configProperty.setValue(get(key));
        }
```

null value is fine for the ServerConfigurationChecker. 
We will first generate a Map<PropertyKey, Map<Value, List<Host:Port>> mConfMap. In this map, null will work as a normal key in the nested map and have all the nodes addresses which have null default values. And then we just check how many values each property key have.

When testing using a new multi process cluster with 2 workers, one has null default value, on has set value, I encounter a NullPointerException and is now working on finding the causes.

```
2018-05-29 21:00:33,456 ERROR TThreadPoolServer - Error occurred during processing of message.
java.lang.NullPointerException
        at org.apache.thrift.protocol.TBinaryProtocol.writeString(TBinaryProtocol.java:200)
        at alluxio.thrift.InconsistentProperty$InconsistentPropertyStandardScheme.write(InconsistentProperty.java:503)
        at alluxio.thrift.InconsistentProperty$InconsistentPropertyStandardScheme.write(InconsistentProperty.java:427)
        at alluxio.thrift.InconsistentProperty.write(InconsistentProperty.java:373)
        at alluxio.thrift.ConfigCheckReport$ConfigCheckReportStandardScheme.write(ConfigCheckReport.java:661)
        at alluxio.thrift.ConfigCheckReport$ConfigCheckReportStandardScheme.write(ConfigCheckReport.java:553)
        at alluxio.thrift.ConfigCheckReport.write(ConfigCheckReport.java:491)
        at alluxio.thrift.GetConfigReportTResponse$GetConfigReportTResponseStandardScheme.write(GetConfigReportTResponse.java:360)
at alluxio.thrift.GetConfigReportTResponse$GetConfigReportTResponseStandardScheme.write(GetConfigReportTResponse.java:322)
        at alluxio.thrift.GetConfigReportTResponse.write(GetConfigReportTResponse.java:273)
        at alluxio.thrift.MetaMasterClientService$getConfigReport_result$getConfigReport_resultStandardScheme.write(MetaMasterClientService.java:4055)
        at alluxio.thrift.MetaMasterClientService$getConfigReport_result$getConfigReport_resultStandardScheme.write(MetaMasterClientService.java:4008)
        at alluxio.thrift.MetaMasterClientService$getConfigReport_result.write(MetaMasterClientService.java:3951)
```
Thanks!

Reply:
@aaudiber Thanks for the suggestions!

It seems that the previous issue happens because in InconsistentProperty, we have mValues which might have null key.  The null keys has problem when write as string.

InconsistentProperty.java
```
/** The name of the property that has errors/warnings. */
  private String mName = "";
  /** Record the values and corresponding hostnames. */
  private Map<String, List<String>> mValues = new HashMap<>();

```

I still think we have to use a string to represent the null value at least in InconsistentProperty and have not come up with better ideas.

Reply:
Thrift doesn't support setting a string to `null`, but does allow the string value to be left unset, which could be used the same way.

Reply:
Then we can change:
```
struct InconsistentProperty {
  1: string name
  2: map<string, list<string>> values
}
```
to 
```
struct InconsistentProperty {
  1: string name
  2: list<InconsistentValue> values
}

struct InconsistentValue {
  1: string value
  2: list<string> addresses
}
```
to support null value.

Reply:
oh sorry, I thought it was already like that. +1, it's good to have the thrift structure mirror the way we use it in java code